Switching docs to use new standard "sync" APIs (#110)
OfficeDev/office-scripts-docs: docs/develop/excel-async-model.md

---
title: 'Using the async Office Scripts APIs in performance-critical scenarios'
description: 'A primer on the Office Scripts async APIs and how to use the load/sync pattern to maximize script performance.'
ms.date: 05/22/2020
localization_priority: Normal
---

# Using the async Office Scripts APIs in performance-critical scenarios

> This article will teach you how to write scripts using the async APIs. These APIs let your script control the data synchronization between the script and the workbook. This gives you maximum control over the network calls to and from the workbook, which are the primary cause of performance issues.

Author: Good call. I'll frame it more like a branding thing, as we do in the ref docs (as per design's suggestion).

> [!IMPORTANT]

> > The async model is significantly more complicated than the standard Office Scripts APIs. We highly recommend following the guidance in [Improve the performance of your Office Scripts](web-client-performance.md) before switching to the async APIs.

Reviewer: Given the situation with this, can you say we recommend using this only if you have performance related concerns or have already been using code that does this? Otherwise, the other one

Author: Current usage is pretty low, so I wanted to frame everything to be forward-facing.

- Additional `Excel.run` calls could be made later in the script. You could also pull them into separate functions called by `main`.

```TypeScript
async function main(workbook: ExcelScript.Workbook) {

    // Standard scripting code...

    // A standard async API block.
    await Excel.run(async (context: Excel.RequestContext) => {
        // Async API code...
    });

    // More standard scripting code...

    // An async API block with a returned value.
    let x = await Excel.run(async (context: Excel.RequestContext) => {
        // Async API code...
        return myValue;
    });
}
```

## Context

The `main` function accepts an `Excel.RequestContext` parameter, named `context`. Think of `context` as the bridge between your script and the workbook. Your script accesses the workbook with the `context` object and uses that `context` to send data back and forth.

The `context` object is necessary because the script and Excel are running in different processes and locations. The script will need to make changes to or query data from the workbook in the cloud. The `context` object manages those transactions.

> By default, Office Scripts handle the interactions between your script and the workbook automatically. While the process is optimized, the standard Office Scripts APIs may synchronize the workbook with your script more than necessary, such as during looped read operations. You may be able to manage these workbook-script data transactions more efficiently with the async APIs, since you know when it is necessary to update data.

Reviewer: ```suggestion
By default, Office Scripts handle the interactions between your script and the workbook automatically. If your code doesn't have performance-related issues with synchronization, such as a looped read operations, then we recommend using [the synchronous API instead.](LINK)
```
If ENG can't think of why then I don't think we should write about that either, maybe just keep steering people away

- Read data from the workbook (following a `load` operation or method that returns a [ClientResult](/javascript/api/office-scripts/excel/excel.clientresult)).
- Write data to the workbook (usually because the script has finished).

The following image shows an example control flow between the script and workbook:

![A diagram showing read and write operations going to the workbook from the script.](../images/load-sync.png)

### Sync

Whenever your async script needs to read data from or write data to the workbook, call the `RequestContext.sync` method as shown here:

```TypeScript
await context.sync();
```

> [!NOTE]
> `context.sync()` is implicitly called when a script ends.

After the `sync` operation completes, the workbook updates to reflect any write operations that script has specified. A write operation is setting any property on a Excel object (e.g. `range.format.fill.color = "red"`) or calling a method that changes a property (e.g., `range.format.autoFitColumns()`). The `sync` operation also reads any values from the workbook that the script requested by using a `load` operation or a method that returns a `ClientResult` (as discussed in the next sections).

> Synchronizing your script with the workbook can take time, depending on your network. You should minimize the number of `sync` calls to help your script run fast. Otherwise, you may as well use the standard, synchronous APIs.

Reviewer: ```suggestion
Synchronizing your script with the workbook can take time, depending on your network. Minimize the number of `sync` calls to help your script run fast, or simply use the standard, synchronous APIs.
```

### Load

An async script must load data from the workbook before reading it. However, loading data from the entire workbook would greatly reduce the script's speed. The `load` method lets your script specifically state what data should be retrieved from the workbook.

> The `load` method is available on every Excel object. Your script must load an object's properties before it can read them. Not doing so will result in an error.

Reviewer: ```suggestion
The `load` method is available on every Excel object. Your script must load an object's properties before it can read them. Not doing so results in an error.
```

The following examples use a `Range` object to show the three ways the `load` method can be used to load data.

|Intent |Example Command | Effect |
|:--|:--|:--|
|Load one property |`myRange.load("values");` | Loads a single property, in this case the two-dimensional array of values in this range. |
|Load multiple properties |`myRange.load("values, rowCount, columnCount");`| Loads all the properties from a comma-delimited list, in this example the values, row count, and column count. |

> |Load everything | `myRange.load();`|Loads all the properties on the range. This is not a recommended solution, since it will slow down your script by getting unnecessary data. You should only use this while testing your script or if you need every property from the object. |

Reviewer: ```suggestion
|Load everything | `myRange.load();`|Loads all the properties on the range. This isn't a recommended solution, since it will slow down your script by getting unnecessary data. Only use this while testing your script or if you need every property from the object. |
```
